zerver/decorator.py

```diff
@@ -748,12 +755,13 @@ def wrapped_func(request: HttpRequest, *args: object, **kwargs: object) -> HttpR
 
             user = request.user
 
-            if isinstance(user, AnonymousUser) or (settings.ZILENCER_ENABLED and
-                                                   isinstance(user, RemoteZulipServer)):
+            if isinstance(user, AnonymousUser):
                 ip_addr = request.META["REMOTE_ADDR"]
                 assert ip_addr
                 rate_limit_ip(request, ip_addr, domain='api_by_ip')
                 return func(request, *args, **kwargs)
+            elif settings.ZILENCER_ENABLED and isinstance(user, RemoteZulipServer):
+                rate_limit_remote_server(request, user, domain='api_by_remote_server')
```

Comment:
What about the other side of this -- are remote servers going to deal gracefully with getting 429's?  Will they just drop the push notification?  Does this implicitly limit the number of push notifications they're going to be allowed to have us send, per minute?  If so, that seems worth documenting.

Reply:
We were actually *accidentally* applying rate-limiting to remote servers for a long time, prior to https://github.com/zulip/zulip/commit/934bdb9651a219edb6b7581dfdd48e889d652eb5 - because the `rate_limit_user` call would actually accept a `RemoteZulipServer` in the `user` argument and happily convert it to a redis key and run its rate limiting on that, because python duck typing :laughing: So for that reason, it may be a reasonable to assume that if rate limiting wasn't an issue back then, then it shouldn't after adding this code.

Looking at the code, it seems like t he `429` will go here in `send_to_push_bouncer` and throw a `JsonableError`, which I think will get logged and drop the push notification from the queue worker?
```
    elif res.status_code >= 400:
        # If JSON parsing errors, just let that exception happen
        result_dict = orjson.loads(res.content)
        msg = result_dict["msg"]
        if "code" in result_dict and result_dict["code"] == "INVALID_ZULIP_SERVER":
            # Invalid Zulip server credentials should email this server's admins
            raise PushNotificationBouncerException(
                _("Push notifications bouncer error: {}").format(msg)
            )
        else:
            # But most other errors coming from the push bouncer
            # server are client errors (e.g. never-registered token)
            # and should be handled as such.
            raise JsonableError(msg)
```

Could be worth:
1. Perhaps throwing `PushNotificationBouncerRetryLaterError` instead to do a retry? Though something slightly more fancy may be needed to communicate to the queue worker like a "retry no earlier than in X seconds"  kind of exception. Thoughts on this?
2. Documenting the limit  as you mentioned - because `handle_push_notification` does a request to the bouncer to send the notification, so that's gonna get rate limited.

Reply:
RabbitMQ has some native ability to do [delayed messages](https://www.cloudamqp.com/docs/delayed-messages.html), which we could potentially be making use of (here and in other places) -- but I think that might get us into the world of weird out-of-order deliveries, since that'd requeue it to the end.  The queue worker should ideally NAK the head of queue message it's working on and sleep for the relevant period of time, so that it gets to re-dequeue the same message.

I think we'll need to document the limit regardless, since doing the above just means that our rate-limiting isn't lossy :). Can you add some documentation with this series that notes this?

Reply:
Documentation added; I'll edit it a bit while merging.

We should open a follow-up issue for making remote Zulip servers respect rate limits.